source/common/config/delta_subscription_impl.cc

```diff
@@ -1,106 +1,89 @@
+    : context_(std::move(context)), type_url_(type_url), callbacks_(callbacks), stats_(stats),
+      init_fetch_timeout_(init_fetch_timeout), is_aggregated_(is_aggregated) {}
 
-void DeltaSubscriptionImpl::pause() { state_->pause(); }
-void DeltaSubscriptionImpl::resume() {
-  state_->resume();
-  trySendDiscoveryRequests();
+DeltaSubscriptionImpl::~DeltaSubscriptionImpl() {
+  if (watch_) {
+    context_->removeWatch(type_url_, watch_);
+  }
 }
 
-// Config::Subscription
-void DeltaSubscriptionImpl::start(const std::set<std::string>& resource_names) {
-  state_ = std::make_unique<DeltaSubscriptionState>(
-      type_url_, resource_names, callbacks_, local_info_, init_fetch_timeout_, dispatcher_, stats_);
-  grpc_stream_.establishNewStream();
-  updateResources(resource_names);
-}
+void DeltaSubscriptionImpl::pause() { context_->pause(type_url_); }
 
-void DeltaSubscriptionImpl::updateResources(const std::set<std::string>& update_to_these_names) {
-  state_->updateResourceInterest(update_to_these_names);
-  // Tell the server about our new interests, if there are any.
-  trySendDiscoveryRequests();
-}
+void DeltaSubscriptionImpl::resume() { context_->resume(type_url_); }
 
-// Config::GrpcStreamCallbacks
-void DeltaSubscriptionImpl::onStreamEstablished() {
-  state_->markStreamFresh();
-  trySendDiscoveryRequests();
+// Config::DeltaSubscription
+void DeltaSubscriptionImpl::start(const std::set<std::string>& resources) {
+  // ADS initial request batching relies on the users of the GrpcMux *not* calling start on it,
+  // whereas non-ADS xDS users must call it themselves.
+  if (!is_aggregated_) {
+    context_->start();
+  }
+  watch_ = context_->addOrUpdateWatch(type_url_, watch_, resources, *this, init_fetch_timeout_);
+  stats_.update_attempt_.inc();
 }
 
-void DeltaSubscriptionImpl::onEstablishmentFailure() { state_->handleEstablishmentFailure(); }
-
-void DeltaSubscriptionImpl::onDiscoveryResponse(
-    std::unique_ptr<envoy::api::v2::DeltaDiscoveryResponse>&& message) {
-  ENVOY_LOG(debug, "Received gRPC message for {} at version {}", type_url_,
-            message->system_version_info());
-  kickOffAck(state_->handleResponse(*message));
+void DeltaSubscriptionImpl::updateResourceInterest(
+    const std::set<std::string>& update_to_these_names) {
+  watch_ = context_->addOrUpdateWatch(type_url_, watch_, update_to_these_names, *this,
+                                      init_fetch_timeout_);
+  stats_.update_attempt_.inc();
 }
 
-void DeltaSubscriptionImpl::onWriteable() { trySendDiscoveryRequests(); }
-
-void DeltaSubscriptionImpl::kickOffAck(UpdateAck ack) {
-  ack_queue_.push(ack);
-  trySendDiscoveryRequests();
+// Config::SubscriptionCallbacks
+void DeltaSubscriptionImpl::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const std::string& version_info) {
+  callbacks_.onConfigUpdate(resources, version_info);
+  stats_.update_success_.inc();
+  stats_.update_attempt_.inc();
```

Comment:
Good question. This exact ordering is ultimately carried over from the old code... and for a second I thought you had found a bug in it!

What actually happens is the throw that can happen within that onConfigUpdate will get caught by the caller of this onConfigUpdate, who will then switch over into onConfigUpdateFailed, which has its own update_attempt.inc().

I think it's safe to call this unacceptably subtle behavior. It's not possible to make it as completely clean as I would like given the current logic, but I have at least changed it so that we're no longer relying on a throw->catch->otherFunction to update the stat. However, looking more closely, I actually don't think the current logic gives the most accurate value for this stat, so I left TODOs.